### PR TITLE
Add ReduceL2Grad and ClipGrad

### PR DESCRIPTION
**Pull request by iK1D**
Add ReduceL2Grad and ClipGrad, which is required by some customer models using nn.functional.normalize.

### CODE REVIEW

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -1444,5 +1476,39 @@ IMPLEMENT_GRADIENT_BUILDER(GetTopKGradient) {
...
+  bool has_i1 = false, has_i2 = false;
+  ArgDef intermediate_arg_def = ArgDef("");
+  // Gradients not defined on min and max, so we return the subgradient 1 for these cases.
+  if (numInputs >= 2 && !(I(1) == ArgDef(""))) {
+    has_i1 = true;
+    intermediate_arg_def = IA("Masked_Min");
+    output.emplace_back(NodeDef("GreaterOrEqual", {I(0), I(1)}, {intermediate_arg_def}));
+  }
+
+  if (numInputs >= 3 && !(I(2) == ArgDef(""))) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1491]: I(2).Exist()

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 2]
```diff
@@ -1444,5 +1476,39 @@ IMPLEMENT_GRADIENT_BUILDER(GetTopKGradient) {
               {MakeAttribute("axis", axis)})};
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetClipGradient) {
+  std::vector<NodeDef> output;
+  size_t numInputs = GetSrcNodeInputSize();
+  bool has_i1 = false, has_i2 = false;
+  ArgDef intermediate_arg_def = ArgDef("");
+  // Gradients not defined on min and max, so we return the subgradient 1 for these cases.
+  if (numInputs >= 2 && !(I(1) == ArgDef(""))) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1485]: I(1).Exist()

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 3]
```diff
@@ -1444,5 +1476,39 @@ IMPLEMENT_GRADIENT_BUILDER(GetTopKGradient) {
...
+    }
+  }
+
+  if (!has_i1 && !has_i2) {
+    output.emplace_back(NodeDef("Identity", {GO(0)}, {GI(0)}));
+  } else {
+    NodeDef zero_constant_node = ZeroConstantNode(IElemType(0));
+    ArgDef ZERO = zero_constant_node.output_args[0];
+    output.emplace_back(zero_constant_node);
+    output.emplace_back(NodeDef("Where", {intermediate_arg_def, GO(0), ZERO}, {GI(0)}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1507]: Just another Idea, 
dX  = dY * cast(mask, type= IElemType(0))

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 4]
```diff
@@ -1101,6 +1101,38 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceLogSumExpGradient) {
...
+  ArgDef dy_arg_def = GO(0);
+  if (!keepdims && attributes.find("axes") != attributes.end()) {
+    std::vector<int64_t> axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+    y_arg_def = IA("Unsqueezed_Y");
+    dy_arg_def = IA("Unsqueezed_dY");
+    result.emplace_back(NodeDef("Unsqueeze", {O(0)}, {y_arg_def}, {MakeAttribute("axes", axes_values)}));
+    result.emplace_back(NodeDef("Unsqueeze", {GO(0)}, {dy_arg_def}, {MakeAttribute("axes", axes_values)}));
+  }
+
+  result.emplace_back(NodeDef("Div", {dy_arg_def, y_arg_def}, {IA("Scaled_dY")}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1123]: Swap order between "Div" and "Unsqueezes"? 
so you can potentially perform Div on a smaller tensor and save one Unsqueeze node.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 5]
```diff
@@ -1101,6 +1101,38 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceLogSumExpGradient) {
...
+
+  // Handle 0 elements in Y.
+  NodeDef zero_constant_node = ZeroConstantNode(IElemType(0));
+  ArgDef ZERO = zero_constant_node.output_args[0];
+  result.push_back(zero_constant_node);
+  result.emplace_back(NodeDef("Equal", {y_arg_def, ZERO}, {IA("Masked_Y")}));
+  result.emplace_back(NodeDef("Where", {IA("Masked_Y"), ZERO, IA("Scaled_dY")}, {IA("Masked_Scaled_dY")}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1130]: Cast might be faster then Where op here.

Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1130]: When y=0, the scaled_dy=NaN, scaled_dy * cast(mask) on these elements is NaN * 0, which is not 0 I think.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 4 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1123]: The element size is same before and after "Unsqueeze", but yes we can save one "Unsqueeze".

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 5 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder.cc]: good point.

**Review by SherlockNoMad — APPROVED** [review]
